proposals/scripting-support.md

```diff
@@ -0,0 +1,434 @@
+  - `(scriptObject, compilationConfig, environment) -> Unit`
+  - uses `compilationConfig` and `environment` for "feeding" the script with the arguments
+  - predefined platform-specific executors available, but usually provided by the scripting host
+  - possible executors
+    - JSR-223
+    - IDEA REPL
+    - Jupyter
+    - Gradle
+    - with specific coroutines context
+    - ...
+- **Preprocessor** - receives the script text before compilation and extracts the `compilationConfig`
+  - `(scriptSource) -> compilationConfig`
+  - *previously called DependenciesResolver*
+  - have access to specific compiler services to:
+    - extract script (file-level) annotations
+    - extract parts of the script (*source sections plugin*)
+    - (potentially) perform script lexing for advanced script analysis
+  - predefined platform-specific preprocessors are available for standard cases, custom one could be provided by the 
+    scripting host
+- **Selector** - receives the script file and/or text before further analysis to determine whether the script belongs
+  to this scripting host and the particular set of components
+  - `(scriptSource) -> Boolean`
+  - required for IDE support
+  - could be combined with the preprocessor, since it may require the same text analysing services 
+  - default and simple implementations are provided
+
+### Script object
+
+Script compiled into the following class
+
+```
+class Script(templateParams..., val implicitReceivers[], additionalArgs: Map<String, Any?>): ScriptDefinition(definitionClassParams...) {
+    
+    val val1: V1 by initOnce // for all vals/vars defined in the script body
+    
+    fun fn1(...): R1 {} // for all funcs defined in the script body
+    
+    class Cl1(...) {} // for all classes/objects defined in the script body
+    
+    var param1: P1 = additionalArgs["param1"] // for all additionalParams, the types are defined by `compilationConfig` extracted by Preprocessor 
+    
+    [suspend] fun <SAM name>(lambdaParams...): RetVal {
+        with(implicitReceivers) {
+            ...
+        }
+    }
+}
+```
+
+In the REPL mode `implicitReceivers` will contain all previous script lines (the order is significant).
+
+The `implicitReceivers` may also contain compiled external scripts objects.
+
+*(The `with (implicitReceivers)` wrapping is needed in all methods and initializers defined in the class.)*
+
+The additional parameters may include params extracted by preprocessor (`@file:param("name", String::class)`) or could 
+be used for bindings mapping in JSR-223-like implementations ([KT-18917](https://youtrack.jetbrains.com/issue/KT-18917)).  
+
+### Script definition
+
+**Script Definition** is a configuration entity that combines platform, preprocessor, and selector in one entity 
+for simplified discovery and configuration. 
+
+The definition is a kotlin class declaration annotated with specific annotations. The class should be defined in the SAM 
```

Comment:
Does it have to be abstract? We could use an annotation like `@ScriptBody` so that the script is generated into an override, no matter whether the member was abstract or not. Must not be final though